File: tests/fixtures/python/sample-repo/src/sample_repo/io_ops.py
"""IO-effect patterns."""
from __future__ import annotations

import os
import shutil
import sys
from pathlib import Path


def write_log(path: Path, msg: str) -> None:
    # expect: IO=definite
    with open(path, "a", encoding="utf-8") as f:
        f.write(msg + "\n")


def read_config(path: Path) -> str:
    # expect: IO=probable
    # Reading is IO in our taxonomy, but read mode is `probable` (the rules file says so).
    return path.read_text(encoding="utf-8")


def emit(line: str) -> None:
    # expect: IO=definite
    print(line)


def emit_err(line: str) -> None:
    # expect: IO=definite
    sys.stderr.write(line + "\n")


def remove_temp(path: Path) -> None:
    # expect: IO=definite
    os.remove(path)


def archive(src: Path, dst: Path) -> None:
    # expect: IO=definite
    shutil.copy(src, dst)


def parse_file(path: Path) -> int:
    # expect: IO=possible
    # Pure body — name pattern only. `path` parameter triggers IO=possible heuristic.
    return len(str(path))

File: tests/fixtures/rust/sample-repo/src/io_ops.rs
// IO-effect patterns.

use std::fs::{self, File, OpenOptions};
use std::io::{BufRead, BufReader, Write};
use std::path::Path;

pub fn write_log(path: &Path, msg: &str) -> std::io::Result<()> {
    // expect: IO=definite
    let mut f = OpenOptions::new().create(true).append(true).open(path)?;
    writeln!(f, "{msg}")
}

pub fn read_config(path: &Path) -> std::io::Result<String> {
    // expect: IO=definite
    fs::read_to_string(path)
}

pub fn emit(line: &str) {
    // expect: IO=definite
    println!("{line}");
}

pub fn emit_err(line: &str) {
    // expect: IO=definite
    eprintln!("{line}");
}

pub fn remove_temp(path: &Path) -> std::io::Result<()> {
    // expect: IO=definite
    fs::remove_file(path)
}

pub fn count_lines(path: &Path) -> std::io::Result<usize> {
    // expect: IO=definite
    let f = File::open(path)?;
    let r = BufReader::new(f);
    Ok(r.lines().count())
}

pub fn dump_path(path: &Path) -> String {
    // expect: IO=probable
    // Takes &Path but does not actually read; rule is a Path-parameter heuristic at probable.
    path.display().to_string()
}
